presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractDecimalSelectiveStreamReader.java

```diff
@@ -0,0 +1,324 @@
+        }
+
+        readOffset = offset + streamPosition;
+        return outputPositionCount;
+    }
+
+    protected int readAllNulls(int[] positions, int positionCount)
+            throws IOException
+    {
+        presentStream.skip(positions[positionCount - 1]);
+
+        if (nonDeterministicFilter) {
+            outputPositionCount = 0;
+            for (int i = 0; i < positionCount; i++) {
+                if (filter.testNull()) {
+                    outputPositionCount++;
+                }
+                else {
+                    outputPositionCount -= filter.getPrecedingPositionsToFail();
+                    i += filter.getSucceedingPositionsToFail();
+                }
+            }
+        }
+        else if (nullsAllowed) {
+            outputPositionCount = positionCount;
+        }
+        else {
+            outputPositionCount = 0;
+        }
+
+        allNulls = true;
+        return positions[positionCount - 1] + 1;
+    }
+
+    abstract int readNoFilter(int[] positions, int position)
+            throws IOException;
+
+    abstract int readWithFilter(int[] positions, int position)
+            throws IOException;
+
+    abstract void ensureValuesCapacity(int capacity, boolean recordNulls);
+
+    protected void skip(int items)
+            throws IOException
+    {
+        if (dataStream == null) {
+            presentStream.skip(items);
+        }
+        else if (presentStream != null) {
+            int dataToSkip = presentStream.countBitsSet(items);
+            dataStream.skip(dataToSkip);
+            scaleStream.skip(dataToSkip);
+        }
+        else {
+            dataStream.skip(items);
+            scaleStream.skip(items);
+        }
+    }
+
+    @Override
+    public int[] getReadPositions()
+    {
+        return outputPositions;
+    }
+
+    @Override
+    public void throwAnyError(int[] positions, int positionCount)
+    {
+    }
+
+    private BlockLease newLease(Block block)
+    {
+        valuesInUse = true;
+        return ClosingBlockLease.newLease(block, () -> valuesInUse = false);
+    }
+
+    @Override
+    public Block getBlock(int[] positions, int positionCount)
+    {
+        checkArgument(outputPositionCount > 0, "outputPositionCount must be greater than zero");
+        checkState(outputRequired, "This stream reader doesn't produce output");
+        checkState(positionCount <= outputPositionCount, "Not enough values");
+        checkState(!valuesInUse, "BlockLease hasn't been closed yet");
+
+        if (allNulls) {
+            return new RunLengthEncodedBlock(nullBlock, outputPositionCount);
+        }
+
+        boolean includeNulls = nullsAllowed && presentStream != null;
+
+        if (positionCount == outputPositionCount) {
+            Block block = makeBlock(positionCount, nullsAllowed, nulls, values);
+            nulls = null;
+            values = null;
+            return block;
+        }
+
+        long[] valuesCopy = createValuesCopy(positionCount);
```

Comment:
consider replacing `createValuesCopy` and `ensureValuesCapacity` abstract methods with a member variable `valuesPerPosition`; let the subclasses provide the value (2 or 1) to the constructor.

`long[] valuesCopy = new long[valuesPerPosition * positionCount];`

otherwise, `createValuesCopy` needs a better name; currently it suggests that the method is making a copy, but it is only allocating memory

Reply:
good idea